providers/domainnameshop/api.go

```diff
@@ -0,0 +1,248 @@
+func (api *domainNameShopProvider) getDNS(domainName string) ([]domainNameShopRecord, error) {
+	domainID, err := api.getDomainID(domainName)
+	if err != nil {
+		return nil, err
+	}
+
+	client := &http.Client{}
+	req, err := http.NewRequest(http.MethodGet, rootAPIURI+"/domains/"+domainID+"/dns", nil)
+	if err != nil {
+		// handle error
+		return nil, err
+	}
+	req.SetBasicAuth(api.Token, api.Secret)
+	resp, err := client.Do(req)
+
+	if err != nil {
+		// handle error
+		return nil, err
+	}
+	defer resp.Body.Close()
+	domainResponse := make([]domainNameShopRecord, 1)
+	err = json.NewDecoder(resp.Body).Decode(&domainResponse)
+	if err != nil {
+		return nil, err
+	}
+
+	for i := range domainResponse {
+		// Fix priority
+		record := &domainResponse[i]
+		priority, err := strconv.ParseUint(record.Priority, 10, 16)
+		if err != nil {
+			record.ActualPriority = 0
+		}
+		record.ActualPriority = uint16(priority)
+
+		port, err := strconv.ParseUint(record.Port, 10, 16)
+		if err != nil {
+			record.ActualPort = 0
+		}
+		record.ActualPort = uint16(port)
+
+		weight, err := strconv.ParseUint(record.Weight, 10, 16)
+		if err != nil {
+			record.ActualWeight = 0
+		}
+		record.ActualWeight = uint16(weight)
+
+		// Fix CAA flags
+		if record.Type == "CAA" {
+			CaaFlag, err := strconv.ParseUint(record.ActualCAAFlag, 10, 8)
+			if err != nil {
+				record.CAAFlag = 0
+			}
+			record.CAAFlag = CaaFlag
+		}
+
+		// Transform data field to punycode if CNAME
+		if record.Type == "CNAME" {
+			punycodeData, err := idna.ToASCII(record.Data)
+			if err != nil {
+				return nil, err
+			}
+			record.Data = punycodeData
+			if !strings.HasSuffix(record.Data, ".") {
+				record.Data += "."
+			}
+		}
+
+		record.TTL = uint16(fixTTL(uint32(record.TTL)))
+
+		// Add domain id
+		(&domainResponse[i]).DomainID = domainID
+	}
+
+	ns, err := api.getNS(domainName)
+	if err != nil {
+		return nil, err
+	}
+
+	// Adds NS as records
+	for _, nameserver := range ns {
+		domainResponse = append(domainResponse, domainNameShopRecord{
+			ID:       0,
+			Host:     "@",
+			TTL:      300,
+			Type:     "NS",
+			Data:     nameserver + ".",
+			DomainID: domainID,
+		})
+	}
+
+	return domainResponse, nil
+}
+
+func (api *domainNameShopProvider) deleteRecord(domainID string, recordID string) error {
+	client := &http.Client{}
+
+	req, err := http.NewRequest(http.MethodDelete, rootAPIURI+"/domains/"+domainID+"/dns/"+recordID, nil)
+	if err != nil {
+		// handle error
+		return err
+	}
+	req.SetBasicAuth(api.Token, api.Secret)
+	resp, err := client.Do(req)
+	if err != nil {
+		// handle error
+		return err
+	}
+
+	switch resp.StatusCode {
+	case 204:
+		//Record is deleted
+		return nil
+	case 403:
+		return fmt.Errorf("not authorized")
+	case 404:
+		return fmt.Errorf("DNS record does not exist")
+	default:
+		return fmt.Errorf("unknown statuscode: %v", resp.StatusCode)
+	}
+}
+
+func (api *domainNameShopProvider) CreateRecord(domainName string, dnsR *domainNameShopRecord) error {
```

Comment:
deleteRecord, CreateRecord and UpdateRecord are very similar.  My recommendation is to put the common code in a separate function.  That function could then do all the error processing. In fact, it could also handle rate-limiting (http code 429) in a single place.